File: Source/DotNET/Backend/Features/FeaturesProvider.cs
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.IO;
using System.Reactive.Subjects;

namespace Dolittle.Vanir.Backend.Features
{

    /// <summary>
    /// Represents an implementation of <see cref="IFeaturesProvider"/>
    /// </summary>
    public class FeaturesProvider : IFeaturesProvider
    {
        static readonly string _featuresPath = Path.Combine(".dolittle","features.json");
        readonly BehaviorSubject<Features> _features = new(new Features());
        readonly IFeaturesParser _featuresParser;

        /// <summary>
        /// Initializes a new instance of <see cref="FeaturesProvider"/>
        /// </summary>
        /// <param name="featuresParser"><see cref="IFeaturesParser"/> for parsing from JSON.</param>
        public FeaturesProvider(IFeaturesParser featuresParser)
        {
            _featuresParser = featuresParser;

            var watcher = new FileSystemWatcher(Path.GetDirectoryName(_featuresPath));
            watcher.Changed += (s, e) => LoadFeatures();
            watcher.Created += (s, e) => LoadFeatures();
            watcher.Deleted += (s, e) => LoadFeatures();
            watcher.NotifyFilter =
                        NotifyFilters.Attributes |
                        NotifyFilters.CreationTime |
                        NotifyFilters.DirectoryName |
                        NotifyFilters.FileName |
                        NotifyFilters.LastAccess |
                        NotifyFilters.LastWrite |
                        NotifyFilters.Security |
                        NotifyFilters.Size;
            watcher.EnableRaisingEvents = true;

            LoadFeatures();
        }

        /// <inheritdoc/>
        public IObservable<Features> Features => _features;

        void LoadFeatures()
        {
            if (!File.Exists(_featuresPath))
            {
                _features.OnNext(new Features());
                return;
            }
            var featuresAsJson = File.ReadAllText(_featuresPath);
            var features = _featuresParser.Parse(featuresAsJson);
            _features.OnNext(new Features(features));
        }
    }
}

File: Source/typescript/backend/features/FeaturesProvider.ts
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import fs from 'fs';

import { Feature, Features, IFeaturesProvider } from '@dolittle/vanir-features';
import { BooleanFeatureToggle } from '@dolittle/vanir-features';
import { injectable } from 'tsyringe';
import { Observable, BehaviorSubject } from 'rxjs';

import chokidar from 'chokidar';
import { FeatureDefinition } from './FeatureDefinition';
import { IFeaturesParser } from '../../features/IFeaturesParser';

const featuresPath = './data/features.json';

/**
 * Represents an implementation of {@link IFeaturesProvider}.
 */
@injectable()
export class FeaturesProvider extends IFeaturesProvider {
    readonly _features: BehaviorSubject<Features> = new BehaviorSubject(new Features());

    constructor(private readonly _parser: IFeaturesParser) {
        super();
        chokidar.watch(featuresPath).on('all', (event, path) => {
            this.loadFeatures();
        });
        this.loadFeatures();
    }

    /** @inheritdoc */
    get features(): Observable<Features> {
        return this._features;
    }

    /** @inheritdoc */
    private loadFeatures(): void {
        if (!fs.existsSync(featuresPath)) {
            this._features.next(new Features());
            return;
        }

        const featuresAsJson = fs.readFileSync(featuresPath).toString();
        const features = this._parser.parse(featuresAsJson);
        this._features.next(features);
    }
}
